### PR TITLE
refactor(experimental): add codecs-core package

### PR DESCRIPTION
**Pull request by lorisleiva**
This PR brings [Umi's core serializers](https://github.com/metaplex-foundation/umi/tree/main/packages/umi-serializers-core) library to Web3.js and renames `Serializer` to `Codec` in order to make room for more granular `Encoder` and `Decoder` types.

### CODE REVIEW

**Comment by lorisleiva** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1630** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1630" target="_blank">Graphite</a>  👈
    * **PR #1631** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1631" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1630?utm_source=stack-comment).

File: packages/codecs-core/package.json [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        "eslint-plugin-sort-keys-fix": "^1.1.2",
+        "jest": "^29.7.0",
+        "jest-environment-jsdom": "^29.6.4",
+        "jest-runner-eslint": "^2.1.0",
+        "jest-runner-prettier": "^1.0.0",
+        "prettier": "^2.8",
+        "test-config": "workspace:*",
+        "tsconfig": "workspace:*",
+        "tsup": "7.2.0",
+        "typescript": "^5.1.6",
```
Comment (steveluscher) [inline packages/codecs-core/package.json:86]:
```suggestion
        "typescript": "^5.2.2",
```

File: packages/codecs-core/src/fix-codec.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ */
+export function fixDecoder<T>(decoder: Decoder<T>, fixedBytes: number, description?: string): Decoder<T> {
+    return {
+        ...fixCodecHelper(decoder, fixedBytes, description),
+        decode: (bytes: Uint8Array, offset = 0) => {
+            assertByteArrayHasEnoughBytesForCodec('fixCodec', fixedBytes, bytes, offset);
+            // Slice the byte array to the fixed size if necessary.
+            if (offset > 0 || bytes.length > fixedBytes) {
+                bytes = bytes.slice(offset, offset + fixedBytes);
+            }
```
Comment (steveluscher) [inline packages/codecs-core/src/fix-codec.ts:43]: 👍🏻

File: packages/codecs-core/src/reverse-codec.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+export function reverseDecoder<T>(decoder: Decoder<T>): Decoder<T> {
+    assertFixedSizeCodec(decoder, 'Cannot reverse a codec of variable size.');
+    return {
+        ...decoder,
+        decode: (bytes: Uint8Array, offset = 0) => {
+            const newBytes = mergeBytes([
+                bytes.slice(0, offset),
+                bytes.slice(offset, offset + decoder.fixedSize).reverse(),
+                bytes.slice(offset + decoder.fixedSize),
```
Comment (steveluscher) [inline packages/codecs-core/src/reverse-codec.ts:28]: Still worth maybe skipping these slices in the case that:

* `offset` is `0`, or
* `offset + bytes.length === fixedSize`

File: packages/codecs-core/src/__tests__/fix-codec-test.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        const b = (s: string) => base16.encode(s);
+
+        expect(fixEncoder(a1z26, 42).description).toBe('fixed(42, a1z26)');
+        expect(fixEncoder(a1z26, 42, 'my fixed').description).toBe('my fixed');
+        expect(fixEncoder(a1z26, 12).fixedSize).toBe(12);
+        expect(fixEncoder(a1z26, 12).maxSize).toBe(12);
+
+        expect(fixEncoder(a1z26, 10).encode('helloworld')).toStrictEqual(b('08050c0c0f170f120c04'));
+        expect(fixEncoder(a1z26, 5).encode('helloworld')).toStrictEqual(b('08050c0c0f'));
+        expect(fixEncoder(a1z26, 10).encode('hello')).toStrictEqual(b('08050c0c0f0000000000'));
```
Comment (steveluscher) [inline packages/codecs-core/src/__tests__/fix-codec-test.ts:76]: Take a quick hop on over to this comment here: https://github.com/solana-labs/solana-web3.js/pull/1608#discussion_r1340937589

**Review by steveluscher — APPROVED** [review]
Love this!

[yess](https://app.graphite.dev/macros/solana-labs)

File: packages/codecs-core/src/__tests__/fix-codec-test.ts (hunk 4 above)
Comment (lorisleiva) [reply packages/codecs-core/src/__tests__/fix-codec-test.ts]: Thanks! All done. ✅

File: packages/codecs-core/package.json (hunk 1 above)
Comment (lorisleiva) [reply packages/codecs-core/package.json]: Updated the entire PR stack. ✅

**Comment by lorisleiva** [comment]
### Merge Activity

* **Sep 30, 6:16 PM**: @lorisleiva started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1630).

**Pull request merged**